drivers/bluetooth/hci/hci_nxp_setup.c

```diff
@@ -0,0 +1,1178 @@
+								     ? CMD_HDR_LEN
+								     : fw_upload.length);
+					fw_upload.last_offset = fw_upload.offset;
+					first = false;
+				} else {
+					fw_upload.cmd7_change_timeout_len = CMD_HDR_LEN;
+					fw_upload.wait_hdr_sig = false;
+					return 0;
+				}
+			} else {
+				if (retry > 0) {
+					retry--;
+					fw_upload_send_ack(V3_TIMEOUT_ACK);
+				} else {
+					return -ETIME;
+				}
+			}
+		} else {
+			return -ENOTSUP;
+		}
+	}
+
+	return -EINVAL;
+}
+
+typedef struct {
+	uint32_t uartBaudRate;
+	uint32_t uartDivisio;
+	uint32_t uartClkDivisor;
+} uart_baudrate_clkDiv_map_t;
+
+static const uart_baudrate_clkDiv_map_t clk_div_map[] = {
+	{115200U, 16U, 0x0075F6FDU},
+	{1000000U, 2U, 0x00800000U},
+	{3000000U, 1U, 0x00C00000U},
+};
+
+static int fw_upload_change_speed_config(struct change_speed_config *config, uint32_t speed)
+{
+	config->clk_div_addr = 0x7f00008fU;
+	config->uart_clk_div_addr = 0x7f000090U;
+	config->mcr_addr = 0x7f000091U;
+	config->reinit_addr = 0x7f000092U;
+	config->icr_addr = 0x7f000093U;
+	config->fcr_addr = 0x7f000094U;
+
+	config->mcr_val = 0x00000022U;
+	config->reinit_val = 0x00000001U;
+	config->icr_val = 0x000000c7U;
+	config->fcr_val = 0x000000c7U;
+
+	for (int i = 0; i < ARRAY_SIZE(clk_div_map); i++) {
+		if (speed == clk_div_map[i].uartBaudRate) {
+			config->clk_div_val = clk_div_map[i].uartClkDivisor;
+			config->uart_clk_div_val = clk_div_map[i].uartDivisio;
+			return 0;
+		}
+	}
+	return -ENOTSUP;
+}
+
+static uint16_t fw_upload_wait_length(uint8_t flag)
+{
+	uint8_t buffer[4];
+	uint16_t len;
+	uint16_t len_comp;
+	int err;
+	uint8_t ack;
+
+	err = fw_upload_read_data(buffer, sizeof(buffer));
+	if (err < 0) {
+		return 0;
+	}
+
+	len = sys_get_le16(buffer);
+	len_comp = sys_get_le16(buffer);
+
+	if ((len ^ len_comp) == 0xFFFF) {
+		LOG_DBG("remote asks for %d bytes", len);
+
+		/* Successful. Send back the ack. */
+		if ((fw_upload.hdr_sig == V1_HEADER_DATA_REQ) ||
+		    (fw_upload.hdr_sig == V1_START_INDICATION)) {
+			ack = V1_REQUEST_ACK;
+			fw_upload_write_data(&ack, 1);
+			if (fw_upload.hdr_sig == V1_START_INDICATION) {
+				/* Eliminated longjmp(resync, 1); returning restart status */
+				return (uint16_t)V1_START_INDICATION;
+			}
+		}
+	} else {
+		LOG_ERR("remote asks len %d bytes", len);
+		LOG_ERR("remote asks len_comp %d bytes", len_comp);
+		/* Failure due to mismatch. */
+		ack = 0xbf;
+		fw_upload_write_data(&ack, 1);
+		/* Start all over again. */
+		if (flag) {
+			/* Eliminated longjmp(resync, 1); returning restart status */
+			return (uint16_t)V1_START_INDICATION;
+		}
+		len = 0;
+	}
+	return len;
+}
+
+static uint32_t fw_upload_get_payload_length(uint8_t *cmd)
+{
+	uint32_t len;
+
+	len = sys_get_le32(&cmd[8]);
+
+	return len;
+}
+
+static void fw_upload_get_hdr_start(uint8_t *buffer)
+{
+	int err;
+	bool done = false;
+	uint32_t count = 0;
+
+	while (!done) {
+		err = fw_upload_read_data(&fw_upload.hdr_sig, 1);
+		if (err >= 0) {
```

Comment:
Check for negative values, otherwise you might have an infinite loop.

Reply:
Updated